src/Compilers/CSharp/Test/Emit2/Diagnostics/GetDiagnosticsTests.cs

```diff
@@ -775,5 +775,44 @@ void M1()
             var diagnosticsByAnalyzerMap = result.SemanticDiagnostics[syntaxTree];
             Assert.Empty(diagnosticsByAnalyzerMap);
         }
+
+        [Fact]
+        [WorkItem(63923, "https://github.com/dotnet/roslyn/issues/63923")]
+        public async Task TestEqualityForCompilerAnalyzerDiagnosticWithPropertyBag()
+        {
+            var source = @"using System;
+
+public class SomeClass
+{
+    [property: Test]
+    public string Name;
+}
+
+internal class TestAttribute : Attribute
+{
+}
+";
+            var compilation = CreateCompilation(source);
+            var compilationDiagnostics = compilation.GetDiagnostics();
+            compilationDiagnostics.Verify(
+                // (5,6): warning CS0657: 'property' is not a valid attribute location for this declaration. Valid attribute locations for this declaration are 'field'. All attributes in this block will be ignored.
+                //     [property: Test]
+                Diagnostic(ErrorCode.WRN_AttributeLocationOnBadDeclaration, "property").WithArguments("property", "field").WithLocation(5, 6));
+            var compilationDiagnostic = compilationDiagnostics.Single();
+
+            var analyzers = ImmutableArray.Create<DiagnosticAnalyzer>(new CSharpCompilerDiagnosticAnalyzer());
+            var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers);
+            var analyzerDiagnostics = await compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync();
+            var analyzerDiagnostic = analyzerDiagnostics.Single();
+
+            // Verify equality for the compiler diagnostic reported from 'CSharpCompilerDiagnosticAnalyzer' with itself.
+            Assert.Equal(analyzerDiagnostic, analyzerDiagnostic);
+
+            // Verify that diagnostic equality check fails when compared with the same compiler diagnostic
+            // fetched from 'compilation.GetDiagnostics()'. Hosts that want to compare compiler diagnostics from
+            // different sources should use custom equality comparer.
+            Assert.NotEqual(analyzerDiagnostic, compilationDiagnostic);
```

Comment:
>analyzerDiagnostic, compilationDiagnostic [](http://example.com/codeflow?start=28&length=41)

How do we know the underlying diagnostics would compare equal for these two?

Would it make sense to make `CompilationAnalyzer.CompilerDiagnostic` an `internal class` so we could test constructing and comparing two instances directly here?

Reply:
Thanks, I added a check to verify ToString representations for both diagnostics are equal.